plugins/parsers/opentsdb/parser.go

```diff
@@ -0,0 +1,127 @@
+package opentsdb
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/parsers"
+)
+
+// Parser encapsulates a OpenTSDB Parser.
+type Parser struct {
+	DefaultTags map[string]string `toml:"-"`
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	var metrics []telegraf.Metric
+
+	scanner := bufio.NewScanner(bytes.NewReader(buf))
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		// delete LF and CR
+		if line[len(line)-1] == '\n' {
+			line = line[:len(line)-1]
+		}
+		if line[len(line)-1] == '\r' {
+			line = line[:len(line)-1]
+		}
+
+		m, err := p.ParseLine(line)
+		if err != nil {
+			return nil, err
+		}
+
+		if m != nil {
+			metrics = append(metrics, m)
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return metrics, nil
+}
+
+// ParseLine performs OpenTSDB parsing of a single line.
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	// Break into fields ("put", name, timestamp, value, tag1, tag2, ..., tagN).
+	fields := strings.Fields(line)
+	if len(fields) == 0 {
+		return nil, nil
+	}
+	if len(fields) < 4 || fields[0] != "put" {
+		return nil, fmt.Errorf("received %q which doesn't have required fields", line)
+	}
+
+	// decode the name and tags
+	measurement := fields[1]
+	tsStr := fields[2]
+	valueStr := fields[3]
+	tagStrs := fields[4:]
+
+	// Parse value.
+	v, err := strconv.ParseFloat(valueStr, 64)
+	if err != nil {
+		return nil, fmt.Errorf(`parsing field %q value failed: %w`, measurement, err)
+	}
+
+	fieldValues := map[string]interface{}{}
+	fieldValues["value"] = v
+
+	// Parse timestamp.
+	ts, err := strconv.ParseInt(tsStr, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf(`parsing field %q time failed: %w`, measurement, err)
+	}
+
+	var timestamp time.Time
+	switch len(tsStr) {
+	case 10:
+		// second resolution
+		timestamp = time.Unix(ts, 0)
+	case 13:
+		// millisecond resolution
+		timestamp = time.UnixMilli(ts)
+	default:
+		return nil, fmt.Errorf(`field %q time: %q time must be 10 or 13 chars`, measurement, tsStr)
+	}
+
+	tags := make(map[string]string)
```

Comment:
We can improve allocation here a bit
```suggestion
	tags := make(map[string]string, len(p.DefaultTags) + len(tagStr))
```

Reply:
`tagStrs`, I assume